### PR TITLE
Ex 4

### PR DESCRIPTION
**Pull request by liortal2**
Just finished exercise 4.
Thanks for the review!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 node_modules
 .idea
 .DS_Store
+.eslintrc
```
Comment (IdanDavidi) [inline .gitignore]: eslintrc should not be part of gitignore - it should be the same for all maintainers and therefore it should be committed

File: server/middleware/cacher.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+module.exports = function cacher(req, res, next) {
+   if (
+      req.body.task
+         .replace(/\s/g, "")
+         .split(",")
+         .every((elem) => !isNaN(elem))
+   ) {
+      let idsToPass = [];
+      const pokemonIDS = req.body.task.replace(/\s/g, "").split(","); // "1, 2, 3" => [1,2,3]
```
Comment (IdanDavidi) [inline server/middleware/cacher.js]: you can reuse `pokemonIDS`
```suggestion
  const potentialPokemonIDS = req.body.task.replace(/\s/g, "").split(","); // "1, 2, 3" => [1,2,3]
   if (
      req.potentialPokemonIDS.every((elem) => !isNaN(elem))
   ) {
      const pokemonIDS = potentialPokemonIDS;
      let idsToPass = [];
```

File: server/middleware/cacher.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      pokemonIDS.forEach((id) => {
+         if (!cachedIDs.includes(id)) {
+            cachedIDs.push(id);
+            idsToPass.push(id);
+         }
+      });
+      if (idsToPass.toString() === "") {
```
Comment (IdanDavidi) [inline server/middleware/cacher.js]: this check is more accurate (and readble)
```suggestion
      if (idsToPass.length === 0) {
```

File: server/middleware/cacher.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+         res.status(200).json(false).send();
```
Comment (IdanDavidi) [inline server/middleware/cacher.js]: this is not what we meant when using cache. what you did is another thing - you block the requests if a user don't ask for new pokemons in the request.
We wanted to "cache" the response from the pokemon API so you won't have to query them every time you get a pokemon - but query them only if you don't have the result in memory. The response from the express server should stay the same no matter if it's stored in the cache or fetched from the pokemon API.

File: server/routes/api.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+const cacher = require("../middleware/cacher");
+const router = express.Router();
+
+router.get("/getAll", (req, res) => {
+   res.status(200).json(taskManager.getTasks());
+});
+
+router.post("/add", cacher, async (req, res) => {
+   await taskManager.add(req.body.task);
+   res.status(200).json(true);
```
Comment (IdanDavidi) [inline server/routes/api.js]: `true` is not a standard json response, prefer returning an empty object `{}` or  object with keys `{ success: true }`

File: server/services/file-manager.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+const fs = require("fs");
+const pokemonClient = require("../clients/pokemon-client.js");
+const POKEMON_FILE_NAME = "./server/pokemon-names.json";
+const TASKS_FILE_NAME = "./server/tasks.json";
+
+class File {
```
Comment (IdanDavidi) [inline server/services/file-manager.js]: Prefer `FileManager`

File: server/services/task-manager.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+         return -1;
+      });
+      if (direction === "down") {
+         sortedTasks = sortedTasks.reverse();
+      }
+      file.clearTasks();
+      sortedTasks.forEach((task) => file.addTask(task));
+   }
+
+   _pokemonTasksHandle(pokemon, pokemonIDS, i) {
```
Comment (IdanDavidi) [inline server/services/task-manager.js]: `i` is not a good variable name. maybe `index`?
anyways, in this case you should have just received the pokemon ID instead - making your method interface much more simple
```suggestion
   _pokemonTasksHandle(pokemon, pokemonID) {
```

File: server/services/task-manager.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+         const pokemonTypes = this._capitalize(pokemonClient.getPokemonTypes(pokemon));
+         const taskToAdd = `Catch ${pokemonName} of type ${pokemonTypes}`;
+         const imageURL = pokemon.sprites.front_default;
+         task = new Task(taskToAdd, pokemonIDS[i], pokemonName, pokemonTypes, imageURL);
+
+         if (file.getAllTasks().includes(taskToAdd)) {
+            task.taskText(
+               `${pokemonName} already exists in your tasks. Please try another Pokemon.`
+            );
+            task.imageURL = [];
```
Comment (IdanDavidi) [inline server/services/task-manager.js]: why do you "clear" the imageURL?
and why do you set it to `[]` instead of `null` for example?

**Review by IdanDavidi — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (liortal2) [reply .gitignore]: Thanks, eventually I did not use eslint and forgot to remove it. Removed now

File: server/routes/api.js (hunk 5 above)
Comment (liortal2) [reply server/routes/api.js]: Changed to  `{ success: true }`

File: server/services/file-manager.js (hunk 6 above)
Comment (liortal2) [reply server/services/file-manager.js]: Agree. Changed.

File: server/services/task-manager.js (hunk 7 above)
Comment (liortal2) [reply server/services/task-manager.js]: Wow thanks, I really didnt need to pass index and just pass pokemonIDS[index]

File: server/services/task-manager.js (hunk 8 above)
Comment (liortal2) [reply server/services/task-manager.js]: Before my tasks file was an array of objects it was an array of just strings (tasks text). It is no longer needed here of course. Thanks, deleted.

**Pull request merged**